### ISSUE TITLE
"value v1 not found on operand stack" error during `stackify` pass

### ISSUE THREAD

**Issue by greenhat** [0xPolygonMiden/compiler#56]
To reproduce - uncomment `test_func_two_arg!(min, core::cmp::min, u32, u32, u32);` test in `instructions.rs` added in #49 

Stack backtrace:
```
---- rust_masm_tests::instructions::min_u32_u32 stdout ----
thread 'rust_masm_tests::instructions::min_u32_u32' panicked at codegen/masm/src/stackify/emitter.rs:724:14:
value v1 not found on operand stack
stack backtrace:
   0: rust_begin_unwind
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/std/src/panicking.rs:617:5
   1: core::panicking::panic_fmt
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/core/src/panicking.rs:67:14
   2: core::panicking::panic_display
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/core/src/panicking.rs:150:5
   3: core::panicking::panic_str
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/core/src/panicking.rs:134:5
   4: core::option::expect_failed
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/core/src/option.rs:1988:5
   5: core::option::Option<T>::expect
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/core/src/option.rs:898:21
   6: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_stack_dependency
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:722:19
   7: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_node
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:637:21
   8: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_inst
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:962:13
   9: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_inst_dependency
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:794:13
  10: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_node
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:605:36
  11: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_inst
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:962:13
  12: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_node
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:615:25
  13: miden_codegen_masm::stackify::emitter::MasmEmitter::emit_schedule
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:513:13
  14: miden_codegen_masm::stackify::emitter::MasmEmitter::emit
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/emitter.rs:470:13
  15: <miden_codegen_masm::stackify::ConvertHirToMasm<&miden_hir::function::Function> as miden_hir::pass::conversion::ConversionPass>::convert
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/mod.rs:172:13
  16: <miden_codegen_masm::stackify::ConvertHirToMasm<miden_hir::module::Module> as miden_hir::pass::conversion::ConversionPass>::convert
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/mod.rs:123:33
  17: <miden_codegen_masm::stackify::ConvertHirToMasm<miden_hir::program::Program> as miden_hir::pass::conversion::ConversionPass>::convert
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/stackify/mod.rs:82:35
  18: miden_codegen_masm::MasmCompiler::compile
             at /Users/dzadorozhnyi/src/miden-ir/codegen/masm/src/lib.rs:62:12
  19: miden_integration_tests::compiler_test::CompilerTest::ir_masm_program
             at ./src/compiler_test.rs:263:27
  20: miden_integration_tests::compiler_test::CompilerTest::expect_masm
             at ./src/compiler_test.rs:231:23
  21: miden_integration_tests::rust_masm_tests::instructions::min_u32_u32
             at ./src/rust_masm_tests/instructions.rs:106:17
  22: miden_integration_tests::rust_masm_tests::instructions::min_u32_u32::{{closure}}
             at ./src/rust_masm_tests/instructions.rs:95:28
  23: core::ops::function::FnOnce::call_once
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/core/src/ops/function.rs:250:5
  24: core::ops::function::FnOnce::call_once
             at /rustc/c469197b19d53a6c45378568f73c00986b20a5a5/library/core/src/ops/function.rs:250:5
```

When the `min(a, b)` Rust code compiled to the following Wasm:
```wat
(module
  (type (;0;) (func (param i32 i32) (result i32)))
  (func $entrypoint (;0;) (type 0) (param i32 i32) (result i32)
    local.get 0
    local.get 1
    local.get 0
    local.get 1
    i32.lt_u
    select
  )
  (memory (;0;) 16)
  (global $__stack_pointer (;0;) (mut i32) i32.const 1048576)
  (global (;1;) i32 i32.const 1048576)
  (global (;2;) i32 i32.const 1048576)
  (export "memory" (memory 0))
  (export "entrypoint" (func $entrypoint))
  (export "__data_end" (global 1))
  (export "__heap_base" (global 2))
)
```

And translated into the following IR:
```
module noname

const $0 = 0x00100000;

global external @__stack_pointer : i32 = $0 { id = 0 };
global external @gv1 : i32 = $0 { id = 1 };
global external @gv2 : i32 = $0 { id = 2 };

pub fn entrypoint(i32, i32) -> i32 {
block0(v0: i32, v1: i32):
    v3 = cast v0 : u32;
    v4 = cast v1 : u32;
    v5 = lt v3, v4 : i1;
    v6 = cast v5 : i32;
    v7 = neq v6, 0 : i1;
    v8 = select v7, v0, v1 : i32;
    br block1(v8);

block1(v2: i32):
    ret v2;
}
```

**Comment by bitwalker** [comment]
Turns out this is a scheduling issue. I'm going to take a little extra time to rework the scheduling code and try to separate the scheduling from the actual codegen - that should make any future scheduling issues such as this easier to diagnose, and make the code generator significantly easier to maintain (it's basically unmaintainable by anyone but me currently, which is not great).

I originally attempted a smaller patch, but due to the specific nature of the bug, fixing it actually just introduces a different-but-similar issue elsewhere. Rather than play whack-a-mole, I'm going to just revisit it ground-up. I had originally planned to rewrite that code anyway because of how tangled it is, so I guess this as good a time as any.

**Comment by bitwalker** [comment]
Fixed by #83
